### PR TITLE
.tar.xz downloader

### PR DESCRIPTION
**Pull request by Rudloff**
Hello,

As stated in https://github.com/composer/composer/issues/2804, it can be useful to be able to install from XZ tarballs. So here is an XzDownloader class based on GzipDownloader.

Regards,

### CODE REVIEW

**Comment by alcohol** [comment]
Would probably benefit from a test (mark as skipped on Windows perhaps?). Other than that, looks good enough. Not very portable though perhaps.

**Comment by Rudloff** [comment]
I added a simple test based on ZipDownloaderTest. Is this the correct way to do it?

File: src/Composer/Downloader/XzDownloader.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    public function __construct(IOInterface $io, Config $config, EventDispatcher $eventDispatcher = null, Cache $cache = null, ProcessExecutor $process = null)
+    {
+        $this->process = $process ?: new ProcessExecutor($io);
+        parent::__construct($io, $config, $eventDispatcher, $cache);
+    }
+
+    protected function extract($file, $path)
+    {
+
```
Comment (alcohol) [inline src/Composer/Downloader/XzDownloader.php]: whitespace

File: src/Composer/Downloader/XzDownloader.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * @author Pierre Rudloff <contact@rudloff.pro>
+ */
+class XzDownloader extends ArchiveDownloader
+{
+    protected $process;
+
+    public function __construct(IOInterface $io, Config $config, EventDispatcher $eventDispatcher = null, Cache $cache = null, ProcessExecutor $process = null)
+    {
+        $this->process = $process ?: new ProcessExecutor($io);
+        parent::__construct($io, $config, $eventDispatcher, $cache);
```
Comment (alcohol) [inline src/Composer/Downloader/XzDownloader.php]: should have a newline before the parent call

File: src/Composer/Downloader/XzDownloader.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    {
+
+        $command = 'tar -xJf ' . ProcessExecutor::escape($file) . ' -C ' . ProcessExecutor::escape($path);
+
+        if (0 === $this->process->execute($command, $ignoredOutput)) {
+            return;
+        }
+
+        $processError = 'Failed to execute ' . $command . "\n\n" . $this->process->getErrorOutput();
+        throw new \RuntimeException($processError);
```
Comment (alcohol) [inline src/Composer/Downloader/XzDownloader.php]: should have a newline before the throw, not after

File: src/Composer/Downloader/XzDownloader.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getFileName(PackageInterface $package, $path)
+    {
+        return $path.'/'.pathinfo(parse_url($package->getDistUrl(), PHP_URL_PATH), PATHINFO_BASENAME);
```
Comment (alcohol) [inline src/Composer/Downloader/XzDownloader.php]: be consistent in using concatenation with or without whitespace, don't mix things up within the same file/class

Comment (Rudloff) [inline src/Composer/Downloader/XzDownloader.php]: Most of this code was copied from GzipDownloader.php which has the same inconsistencies :wink: 
But you're right, I'll edit my code.

Comment (alcohol) [inline src/Composer/Downloader/XzDownloader.php]: Yeah, copy paste isn't always the best solution ;-)

**Comment by Seldaek** [comment]
Thanks!

**Pull request merged**